### ISSUE TITLE
Implement a smart dice selection algorithm

### ISSUE THREAD

**Issue by Jarvis-Yu** [Jarvis-Yu/Dottore-Genius-Invokation-TCG-Simulator#14]
Currently there's a dice selection algorithm in `dgisim/src/dices.py`, the function name is `basically_satisfy()`. This algorithm is not smart enough to make the best choice on which dices to pay for a card or an action. It should be improved for the better.

If you are not familiar with the rules of Genius Invokation TCG, then please leave a comment below, otherwise you should know how in Genshin the dices are automatically selected for you when you want to do something.

As the current character class doesn't have info on the element type of the character, you cannot directly access it from the game_state, instead you should assume you can access that information and build the algorithm on that basis. So when in the future the information is accessible, the algorithm can adapt to the new character system immediately.

Check [wiki](https://github.com/Jarvis-Yu/Dottore-Genius-Invokation-TCG-Simulator/wiki) for code/commit style guide.

**Comment by Tookser** [comment]
I think I can help with this feature. How much info should I know about this game to implement it?

**Comment by Jarvis-Yu** [comment]
Thanks for contacting. tbh, you don't really read to know a lot. You just need to know the rules about what kind of dices should be chosen for what request.

So in my project there are AbstractDices and ActualDices, AbstractDices are the requirement to be satisfied, which contains types: OMNI, ANY, PYRO, HYDRO, ANEMO, ELECTRO, DENDRO, CRYO, GEO.

OMNI is a special type in AbstractDices, e.g. if the AbstractDices say there needs to be 2 OMNI dices, then any two dices of the same kind can satisfy the requirement.

ANY literally means any dices can be used to satisfy the requirement. (as long as the number of dices are enough)

The other dices are just requesting to be provided with the same number of dices of the same kind.

ActualDices contains all types that AbstractDices have, except ANY.

In ActualDices, an OMNI dice can be treated as any type of dice.

## Lemme give a quick example so you'll know what I mean.

```
AbstractDices {
    OMNI: 2
}
```

can be satisfied by:

```
ActualDices {
    OMNI: 1,
    PYRO: 1,
}
```

or

```
ActualDices {
    HYDRO: 2
}
```

or

```
ActualDices {
    OMNI: 2
}
```

## For ANY dices:

```
AbstractDices {
    ANY: 2
}
```

can be satisfied by:

```
ActualDices {
    GEO: 1,
    ANEMO: 1
}
```

To make the dices selection algorithm smart, as a player has limited dices, the usage of some type of dices should be better avoided. e.g. if the player has a PYRO character on field which uses PYRO dices often, playing a card that requires dices `ANY: 2` should better use dices other than OMNI dices and PYRO dices if you can.

So basically the task is: given the dices you have, the dices the player action needs, and the precedence of dices that you may want to keep, design an algorithm that priorities the usage of low precedence dices.

The precedence is given as type `list[set[Element]]`, e.g.

```
precedence = [
    {Element.OMNI},  # should keep as many as possible
    {Element.DENDRO},  # lower precedence to be kept
    {Element.GEO, Element.ELECTRO},  # even lower precedence to be kept
]  # all Elements not in the precedence have the same lowest priority to be kept
```

If you are gonna implement it, I can help you setup the unit tests if you need my help with it.

Sorry for typing a lot unreviewed, hopefully the content is somewhat organized, if you have further questions, please don't hesitate to ask.

**Comment by Tookser** [comment]
Thank you for detail response. 

Only one question remains about precedence. In your example is it always better to save 3 OMNI, than to save 2 OMNI, no matter how many other dices were saved?

Other thing is performance and number of dice. Does performance matter (I think so), and number of dice (I read about 8 dice).

**Comment by Jarvis-Yu** [comment]
> In your example is it always better to save 3 OMNI, than to save 2 OMNI, no matter how many other dices were saved?

Actually, the primary goal is to save the most OMNI + num of dices having secondary priority. e.g. for the request {OMNI: 2} when you have {OMNI: 1, PYRO: 2, GEO: 1} with precedence [{OMNI}, {PYRO}, {GRO}], you should use {OMNI:1, GEO:1} instead of {PYRO:2}. 

The primary goal is only for the dices with the highest priority and secondary priority. Only OMNI dices are likely to have the highest priority in practice, where there can be up to two types of dices with the secondary priority in the current version of the game I'm trying to build a simulator of.

Satisfying the primary goal, yes, you should keep as many OMNI as possible.

> Other thing is performance and number of dice. Does performance matter (I think so), and number of dice (I read about 8 dice).

Performance wise, it doesn't need to be the fastest algorithm for this ever, but should be as fast as possible (don't sacrifice too much readability to do crazy optimization unless really well commented).

About the data size, there should be <= 16 dices in your hand most of the time, and <= 5 dices in the request. My repo should allow some customization in the future, so let's say the algorithm should perform well when there are less than 32 dices. (it doesn't really make a lot of sense for the user to set the maximum dices number to be too large, so despite we do support that, there's no need for any guarantee on the performance when the number is greater than 32)

---

It's quite late now in my time-zone and my mind isn't very clear at the moment, hopefully I have explained things well. if you have any further questions, please don't hesitate to ask.

**Comment by Tookser** [comment]
(not urgent)

I understood more. 

Yet another question about precedence. 

```
precedence = [
    {Element.OMNI},  # should keep as many as possible
    {Element.DENDRO},  # lower precedence to be kept
    {Element.GEO, Element.ELECTRO},  # even lower precedence to be kept
]  # all Elements not in the precedence have the same lowest priority to be kept
```

1. First we minimize OMNI+DENDRO.
2. (holding OMNI+DENDRO same (minimal)) then we minimize OMNI.
3. (assumption) (holding 1 and 2 same, we minimize) DENDRO
4. (assumption) (holding 1,2,3, we minimize) GEO + ELECTRO.

Is it correct?

---

Sorry for night messages.

**Comment by Jarvis-Yu** [comment]
seems correct, though holding 1 and 2 already implies 3 is satisfied.

I'm thinking if there is a better data structure than my current list[set[Element]], like removing the first row of OMNI dices as it is always gonna be in the first row.

Also, it is also a good idea to priorities the use of dices with less of its kind.

e.g. If precedence doesn't contain Element.DENDRO, Element.HYDRO and Element.ANEMO. For a request `{Element.ANY: 2}` with `{DENDRO: 1, ANEMO: 1, HYDRO: 2}` in your hand, you should return `{DENDRO: 1, ANEMO: 1}` but not `{HYDRO: 2}`, as `{HYDRO: 2}` can potentially satisfy `{OMNI: 2}`.

e.g. with

```
precedence = [
    {Element.OMNI},  # should keep as many as possible
    {Element.DENDRO},  # lower precedence to be kept
    {Element.GEO, Element.ELECTRO},  # even lower precedence to be kept
]  # all Elements not in the precedence have the same lowest priority to be kept
```

and `{GEO: 2, ELECRO: 3}` in your hand, for the request `{ANY: 2}`, you should give `{GEO: 2}`.

if you have`{GEO: 2, ELECTRO: 2}` in your hand, then use only the one with higher priority in another given list, let's say `[PYRO, HYDRO, ANEMO, ELECTRO, DENDRO, CRYO, GEO]` for now. So in this case, return {ELECTRO: 2}. (The reason to not choose an arbitrary one is to make sure the algorithm always gives the same output under the same situations.

**Comment by Tookser** [comment]
Okay. I think I figure out algorithm and now I can code it, but without the

> Also, it is also a good idea to priorities the use of dices with less of its kind.

property. Computational complexity is not so high (most probably O(k), k - number of dice types).

Would it be good try to code it now, or it would be better to think about this improvement?

**Comment by Jarvis-Yu** [comment]
I believe it is good to start right away, and write unit tests along the way. (write tests in `dgisim/tests/test_dices.py`)

The number of dice types is not likely to exceed 8, so you don't need to worry too much about the complexity.

Just saying, if possible, it'd be better if some of these 'properties' can be controlled by some boolean value, but if is really complicating the algorithm more than necessary, you may leave it.

The reason I added the property

> Also, it is also a good idea to priorities the use of dices with less of its kind.

is because, in the game, quite often, playing some cards has requirements like `{OMNI: 3}`, `{OMNI: 2}`, so in most cases, leaving as many dices of the same kind (when not violating the *precedence*) gives players a higher change of being able to play cards like these in the future. Except from one card only, for that card, leaving as many kind of dices as possible (up to 3 kinds) is better than leaving <= 2 kind of dices. In the actual game that I'm trying to build a simulator of, they chose to ignore the special request of this card, and allow players to manually choose which dices they want to consume. It'd be cool if this is an optional condition in the algorithm, but as the actual game itself doesn't support this feature, I don't desperately need it either.

**Comment by Tookser** [comment]
Hello! 

Finally I finished algorithm, but it requires some refactoring. Also I wrote 4 simple tests, on some regular cases, I think should be more. 
Can I use pytest framework for tests? It can run unittests as well, main thing I need from it is parametrisation and fixtures.

**Comment by Jarvis-Yu** [comment]
Thanks for your PR, I'm reviewing the code, may take some time.

**Comment by Jarvis-Yu** [comment]
PR merged
